### PR TITLE
[CALCITE-3898] RelOptPredicateList may generate incorrect map of cons…

### PR DESCRIPTION
**Pull request by jcamachor**
…tant values

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rex/RexUtil.java [hunk 1]
```diff
@@ -454,15 +454,48 @@ private static void decompose(Set<RexNode> set, RexNode rexNode) {
...
     if (name1.getFamily() == name2.getFamily()) {
       switch (name1.getFamily()) {
       case NUMERIC:
-        return name1.compareTo(name2) >= 0;
+        if (SqlTypeUtil.isExactNumeric(type1)
+            && SqlTypeUtil.isExactNumeric(type2)) {
+          int precision1;
+          int scale1;
+          if (name1 == SqlTypeName.DECIMAL) {
+            precision1 = type1.getPrecision();
```
Comment (DonnyZone) [inline core/src/main/java/org/apache/calcite/rex/RexUtil.java:469]: Do we need to consider `JavaType`, which returns an invalid `precision` and `scale`.
It is specially handled in many places, e.g., `RelDataTypeSystem#deriveDecimalPlusType/deriveDecimalMultiplyType...`.

Comment (jcamachor) [reply core/src/main/java/org/apache/calcite/rex/RexUtil.java:469]: Thanks @DonnyZone . I was not aware of that issue, it seems relevant. I have changed the code accordingly.

**Comment by DonnyZone** [comment]
LGTM

**Comment by DonnyZone** [comment]
Hi @danny0405, I noticed the `decimalOf` method in `RelDataTypeFactory` is mainly used for the JavaType, which returns unspecified `precision` and `scale`. One trouble thing is that we always need to cope with the `JavaType` specifically when an issue involves type precision and scale.

Could we implement `getPrecision()` and `getScale()` in `JavaType`?

**Pull request closed without merging**